File: Data Structures & Algorithms/binary-search/submission-1.java
class Solution {
    public int search(int[] nums, int target) {
        return find(nums, target, 0, nums.length - 1);
    }

    private int find(int[] nums, int target, int l, int r) {
        if (l > r) return -1;
        int m = (l + r) / 2;
        int val = nums[m];
        
        if (val == target) return m;
        else if (val > target) return find(nums, target, l, m - 1);
        else return find(nums, target, m + 1, r);
    }
}

File: Data Structures & Algorithms/target-sum/submission-1.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = {}

        def dfs(i, cur):
            if i == len(nums):
                if cur == target:
                    dp[(i, cur)] = 1
                else:
                    dp[(i, cur)] = 0

            if not (i, cur) in dp:
                dp[(i, cur)] = dfs(i + 1, cur - nums[i]) + dfs(i + 1, cur + nums[i])
            
            return dp[(i, cur)]

        return dfs(0, 0)